protocol/binary/writer.go

```diff
@@ -59,63 +58,29 @@ type Writer struct {
 //
 // This Writer must be returned back using ReturnWriter.
 func BorrowWriter(w io.Writer) *Writer {
+	streamWriter := streamWriterPool.Get().(*StreamWriter)
+	streamWriter.writer = w
 	writer := writerPool.Get().(*Writer)
-	writer.writer = w
+	writer.sw = streamWriter
 	return writer
 }
 
 // ReturnWriter returns a previously borrowed Writer back to the system.
 func ReturnWriter(w *Writer) {
-	w.writer = nil
+	w.sw.writer = nil
+	streamWriterPool.Put(w.sw)
+	w.sw = nil
```

Comment:
We should maybe use ReturnStreamWriter here. And for safety, we should do that after setting sw to nil because that drops the last reference we have to it.

```suggestion
	sw := w.sw
	w.sw = nil
	ReturnStreamWriter(sw)
	writerPool.Put(w)
```